data/txHandler.go

```diff
@@ -263,19 +281,41 @@ func (handler *TxHandler) checkAlreadyCommitted(tx *txBacklogMsg) (processingDon
 		logging.Base().Debugf("txPool rejected transaction: %v", err)
 		return true
 	}
+	return false
+}
 
-	// build the transaction verification context
-	latest := handler.ledger.Latest()
-	latestHdr, err := handler.ledger.BlockHdr(latest)
-	if err != nil {
-		logging.Base().Warnf("Could not get header for previous block %v: %v", latest, err)
-		return true
-	}
-	tx.proto = config.Consensus[latestHdr.CurrentProtocol]
-	tx.spec.FeeSink = latestHdr.FeeSink
-	tx.spec.RewardsPool = latestHdr.RewardsPool
+// note: this modifies tx
+func (handler *TxHandler) prepareVerifyContexts(tx *txBacklogMsg, latestHdr bookkeeping.BlockHeader) error {
+	// TODO set verifyContexts correctly
+	tx.verifyContexts = make([]verify.Context, len(tx.unverifiedTxGroup))
+	for i, txn := range tx.unverifiedTxGroup {
+		var fvTime uint64
+		if !txn.Lsig.Blank() {
+			// TODO: move this into some lazy evaluator for the few scripts that actually use `txn FirstValidTime` ?
+			if txn.Txn.FirstValid == 0 {
+				return fmt.Errorf("(%d) LogicSig does not work with FirstValid==0", i)
+			}
+			hdr, err := handler.ledger.BlockHdr(txn.Txn.FirstValid - 1)
+			if err != nil {
+				return fmt.Errorf("(%d) could not fetch BlockHdr for FirstValid-1=%d (current=%d): %s", i, txn.Txn.FirstValid-1, latestHdr.Round, err)
+			}
+			if hdr.TimeStamp < 0 {
+				return fmt.Errorf("(%d) cannot evaluate LogicSig before 1970 at TimeStamp %d", i, hdr.TimeStamp)
+			}
+			fvTime = uint64(hdr.TimeStamp)
+		}
 
-	return false
+		var ctx verify.Context
+		ctx.CurrSpecAddrs.FeeSink = latestHdr.FeeSink
+		ctx.CurrSpecAddrs.RewardsPool = latestHdr.RewardsPool
+		ctx.CurrProto = latestHdr.CurrentProtocol
+		ctx.FirstValidTime = fvTime
```

Comment:
nit: Initializing the Params member like this is very misleading. ( while being semantically correct )
```
ctx := verify.Context{
  params : Params {
      CurrSpecAddrs = ...,
  },
  FirstValidTime: fvTime,
}
```
is much more readable ( at least for me )